README.md
# flutter_verification_code [![Pub](https://img.shields.io/pub/v/flutter_verification_code.svg)](https://pub.dev/packages/flutter_verification_code)

- A Flutter package that help you create a verification input.

- Based on https://github.com/tiny-express/flutter_verification_code_input.

- Added state change in case user delete char after complete fill, take a look an example on [Stackoverflow](https://stackoverflow.com/questions/59005381/how-to-know-when-user-delete-the-input-in-verificationcodeinput-flutter/59006077#59006077)

- With version 1.1.0 supports copy/paste of entire code

After install, in your Dart code, you can use:

```dart
import 'package:flutter_verification_code/flutter_verification_code.dart';
```

## Usage

```dart
  VerificationCode(
    textStyle: TextStyle(fontSize: 20.0, color: Colors.red[900]),
    keyboardType: TextInputType.number,
    underlineColor: Colors.amber, // If this is null it will use primaryColor: Colors.red from Theme
    length: 4,
    cursorColor: Colors.blue, // If this is null it will default to the ambient
    // clearAll is NOT required, you can delete it
    // takes any widget, so you can implement your design
    clearAll: Padding(
      padding: const EdgeInsets.all(8.0),
      child: Text(
        'clear all',
        style: TextStyle(fontSize: 14.0, decoration: TextDecoration.underline, color: Colors.blue[700]),
      ),
    ),
    onCompleted: (String value) {
      setState(() {
        _code = value;
      });
    },
    onEditing: (bool value) {
      setState(() {
        _onEditing = value;
      });
      if (!_onEditing) FocusScope.of(context).unfocus();
    },
  ),
```

```dart
onEditing: (bool value) {
  setState(() {
    _onEditing = value;
  });
},
```

```dart
Center(
  child: (_onEditing != true)
      ? Text('Your code: $_code')
      : Text('Please enter full code'),
),
```

## How to use feature with pasting a code
The example is here https://github.com/awaik/flutter_verification_code/issues/23

Full example is here https://github.com/awaik/flutter_verification_code/tree/master/example

## Showcase

![Image|100x100, 10%](show_case_v3.gif)

![Image|100x100, 10%](show_case_v2.gif)

## Credits

This is a project by [Agoradesk](https://agoradesk.com/), P2P cryptocurrency trading platform.
Created by the team behind LocalMonero, the biggest and most trusted Monero P2P trading platform.




CHANGELOG.md
# 1.1.7

Made by https://github.com/avraham-mayer

- If backspace pressed in empty box, move to previous box

# 1.1.6

Made by https://github.com/ebrahim-2

- Add filled property to activate input background color when fillColor has a value

If true the decoration's container is filled with fillColor.

When InputDecorator.isHovering is true, the hoverColor is also blended into the final fill color.

Typically this field set to true if border is an UnderlineInputBorder.

The decoration's container is the area which is filled if filled is true and bordered per the border. It's the area adjacent to icon and above the widgets that contain helperText, errorText, and counterText.

This property is false by default.

# 1.1.5

- Fixed https://github.com/awaik/flutter_verification_code/issues/33, thank to https://github.com/Xorn4400

# 1.1.4

- Added `margin` and `padding` parameters https://github.com/awaik/flutter_verification_code/issues/35
- Example updated

# 1.1.3+1

- Fixed `isSecure` parameter, thanks to https://github.com/heltonrosa


# 1.1.3

- Added fullBorder property. If true - displays boxes instead of underlined placeholders.
- Added final `Stream<bool>? pasteStream;` property. When it receives true, pastes the
clipboard content. With this property is possible to add button or another functionality
outside this widget, that initiates pasting from clipboard.
Look animated example in the readme.md

# 1.1.2

- Added cursorColor property. If this is null it will default to the ambient.

# 1.1.1

- Added option to move with left and right arrow keys. (Changes made by https://github.com/rajrushilmakkar)

# 1.1.0

- Support copy/paste of entire code. (Changes made by https://github.com/yonatann)

# 1.0.4

- Option to allow only number input from keyboard(flutter web) (Changes made by https://github.com/rajrushilmakkar)

# 1.0.3

- Add fillColor option.

# 1.0.2

- can remove after fill the textfield, made by https://github.com/manofi21
- Add underlineUnfocusedColor to change color of boxes without focus and underlineWidth to change line width of the underline, made by https://github.com/tomaash 

# 1.0.1

- Fixed bug with multiple widgets on the screen https://github.com/awaik/flutter_verification_code/issues/14 - removed static definition.

# 1.0.0+1

- Fixed docs

# 1.0.0

- Null safety and Flutter 2.0

# 0.2.1

- Add option 'underlineColor' - in case underline color is null it will use primaryColor from Theme

# 0.2.0

- Add option 'clearAll' - for clear all user input. https://github.com/awaik/flutter_verification_code/issues/6

```
// clearAll is NOT required, you can delete it
// takes any widget, so you can implement your design
clearAll: Padding(
  padding: const EdgeInsets.all(8.0),
  child: Text(
    'clear all',
    style: TextStyle(
        fontSize: 14.0,
        decoration: TextDecoration.underline,
        color: Colors.blue[700]),
  ),
),
```

# 0.1.6+2

- Change onCompleted, onEditing option to @required for fixing issue https://github.com/awaik/flutter_verification_code/issues/5

# 0.1.6+1

- Removed color from constructor, thanks for finding and fixing https://github.com/almayahi

# 0.1.6

- Thanks for finding and fixing bug to https://github.com/mohammadPezesh1994 - "bug occur when create VerificationCode in stateful widget and after created"
- Fixed bug "Keyboard can not display on iOS". Thank for the issue to https://github.com/thienhaole92

# 0.1.5+1

- Improved README.MD

# 0.1.5

- Fixed bug that was connected with FocusNode() and appeared after big Flutter update.


# 0.1.4

- Improved description in README.md files for package and example.


# 0.1.3

- New show case
- `onEditing` can now be used for getting info about state of code completion
```dart
onEditing: (bool value) {
  setState(() {
    _onEditing = value;
  });
},
```

example/test/widget_test.dart
// This is a basic Flutter widget test.
// To perform an interaction with a widget in your test, use the WidgetTester utility that Flutter
// provides. For example, you can send tap and scroll gestures. You can also use WidgetTester to
// find child widgets in the widget tree, read text, and verify that the values of widget properties
// are correct.

//import 'package:flutter/material.dart';
//import 'package:flutter_test/flutter_test.dart';
//
//import 'package:example/main.dart';
//
void main() {}


test/flutter_verification_code_input_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_verification_code/flutter_verification_code.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
    // expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/src/flutter_verification_code.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class VerificationCode extends StatefulWidget {
  /// is completed
  final ValueChanged<String> onCompleted;

  /// is in process of editing
  final ValueChanged<bool> onEditing;

  /// keyboard type
  final TextInputType keyboardType;

  /// quantity of boxes
  final int length;

  /// cursorColor :)
  final Color? cursorColor;

  /// size of box for code
  final double itemSize;

  /// the color for underline, in case underline color is null it will use primaryColor from Theme
  final Color? underlineColor;

  /// the color for underline when not focused, grey by default
  final Color? underlineUnfocusedColor;

  /// display rectangular border instead of underlined
  final bool fullBorder;

  /// the color for TextField background
  final Color? fillColor;

  /// the line width for underline
  final double? underlineWidth;

  /// style of the input text
  final TextStyle textStyle;

  /// auto focus when screen appears
  final bool autofocus;

  ///takes any widget, display it, when tap on that element - clear all fields
  final Widget? clearAll;

  ///display button for the paste from clipboard functionality
  final Stream<bool>? pasteStream;

  /// to secure the TextField
  final bool isSecure;

  ///accept only digit inputs from keyboard
  final bool digitsOnly;

  /// padding between boxes
  final EdgeInsets margin;

  /// padding inside boxes
  final EdgeInsets? padding;

  const VerificationCode({
    Key? key,
    required this.onCompleted,
    required this.onEditing,
    this.keyboardType = TextInputType.number,
    this.length = 4,
    this.cursorColor,
    this.itemSize = 50,
    this.underlineColor,
    this.underlineUnfocusedColor,
    this.fullBorder = false,
    this.fillColor,
    this.underlineWidth,
    this.textStyle = const TextStyle(fontSize: 25.0),
    this.autofocus = false,
    this.clearAll,
    this.pasteStream,
    this.isSecure = false,
    this.digitsOnly = false,
    this.margin = EdgeInsets.zero,
    this.padding,
  }) : super(key: key);

  @override
  _VerificationCodeState createState() => _VerificationCodeState();
}

class _VerificationCodeState extends State<VerificationCode> {
  final List<FocusNode> _listFocusNode = <FocusNode>[];
  final List<FocusNode> _listFocusNodeKeyListener = <FocusNode>[];
  final List<TextEditingController> _listControllerText =
      <TextEditingController>[];
  final List<String> _code = [];
  int _currentIndex = 0;

  @override
  void initState() {
    _listFocusNode.clear();
    _listFocusNodeKeyListener.clear();
    for (var i = 0; i < widget.length; i++) {
      _listFocusNode.add(FocusNode());
      _listFocusNodeKeyListener.add(FocusNode());
      _listControllerText.add(TextEditingController());
      _code.add('');
    }
    if (widget.pasteStream != null) {
      widget.pasteStream!.listen((val) {
        if (val) {
          _pasteFromClipboard();
        }
      });
    }
    super.initState();
  }

  @override
  void dispose() {
    for (final val in _listControllerText) {
      val.dispose();
    }
    for (final val in _listFocusNode) {
      val.dispose();
    }
    super.dispose();
  }

  String _getInputVerify() {
    String verifyCode = "";
    for (var i = 0; i < widget.length; i++) {
      for (var index = 0; index < _listControllerText[i].text.length; index++) {
        if (_listControllerText[i].text[index] != "") {
          verifyCode += _listControllerText[i].text[index];
        }
      }
    }
    return verifyCode;
  }

  Widget _buildInputItem(int index) {
    final padding =
        widget.padding ?? EdgeInsets.all(((widget.itemSize * 2) / 10));

    final underlinedDecoration = InputDecoration(
      fillColor: widget.fillColor,
      filled: widget.fillColor != null ? true : false,
      enabledBorder: UnderlineInputBorder(
        borderSide: BorderSide(
          color: widget.underlineUnfocusedColor ?? Colors.grey,
          width: widget.underlineWidth ?? 1,
        ),
      ),
      focusedBorder: UnderlineInputBorder(
        borderSide: BorderSide(
          color: widget.underlineColor ?? Theme.of(context).primaryColor,
          width: widget.underlineWidth ?? 1,
        ),
      ),
      counterText: "",
      contentPadding: padding,
      errorMaxLines: 1,
    );

    final fullDecoration = InputDecoration(
      fillColor: widget.fillColor,
      filled: widget.fillColor != null ? true : false,
      enabledBorder: OutlineInputBorder(
        borderSide: BorderSide(
          color: widget.underlineUnfocusedColor ?? Colors.grey,
          width: widget.underlineWidth ?? 1,
        ),
      ),
      focusedBorder: OutlineInputBorder(
        borderSide: BorderSide(
          color: widget.underlineColor ?? Theme.of(context).primaryColor,
          width: widget.underlineWidth ?? 1,
        ),
      ),
      counterText: "",
      contentPadding: padding,
      errorMaxLines: 1,
    );

    return RawKeyboardListener(
      focusNode: _listFocusNodeKeyListener[index],
      onKey: (event) {
        if (event.runtimeType == RawKeyUpEvent) {
          if (event.data.logicalKey == LogicalKeyboardKey.backspace &&
              _listControllerText[index].text.isEmpty) {
            if (index > 0) {
              _listControllerText[index - 1].clear();
            }
            _prev(index);
          }
          if (event.data.logicalKey == LogicalKeyboardKey.arrowLeft) {
            _prev(index);
          } else if (event.data.logicalKey == LogicalKeyboardKey.arrowRight) {
            _next(index);
          }
        }
      },
      child: TextField(
        keyboardType: widget.keyboardType,
        inputFormatters: widget.digitsOnly
            ? <TextInputFormatter>[FilteringTextInputFormatter.digitsOnly]
            : null,
        maxLines: 1,
        maxLength: widget.length - index,
        controller: _listControllerText[index],
        focusNode: _listFocusNode[index],
        showCursor: true,
        cursorColor: widget.cursorColor,
        maxLengthEnforcement: MaxLengthEnforcement.enforced,
        autocorrect: false,
        textAlign: TextAlign.center,
        autofocus: widget.autofocus,
        obscureText: widget.isSecure,
        style: widget.textStyle,
        decoration: widget.fullBorder ? fullDecoration : underlinedDecoration,
        onChanged: (String value) {
          if ((_currentIndex + 1) == widget.length && value.isNotEmpty) {
            widget.onEditing(false);
          } else {
            widget.onEditing(true);
          }

          if (value.isEmpty && index >= 0) {
            _prev(index);
            return;
          }

          if (value.isNotEmpty) {
            String _value = value;
            int _index = index;

            while (_value.isNotEmpty && _index < widget.length) {
              _listControllerText[_index].value = TextEditingValue(
                text: _value[0],
                selection: TextSelection.fromPosition(
                  TextPosition(
                    affinity: TextAffinity.downstream,
                    offset: _listControllerText[_index].text.length,
                  ),
                ),
              );
              _next(_index++);
              _value = _value.substring(1);
            }

            if (_listControllerText[widget.length - 1].value.text.length == 1 &&
                _getInputVerify().length == widget.length) {
              widget.onEditing(false);
              widget.onCompleted(_getInputVerify());
            }
          }
        },
      ),
    );
  }

  void _next(int index) {
    if (index != widget.length - 1) {
      setState(() {
        _currentIndex = index + 1;
      });
      FocusScope.of(context).requestFocus(_listFocusNode[_currentIndex]);
    }
  }

  void _prev(int index) {
    if (index > 0) {
      setState(() {
        if (_listControllerText[index].text.isEmpty) {}
        _currentIndex = index - 1;
      });
      FocusScope.of(context).requestFocus(FocusNode());
      FocusScope.of(context).requestFocus(_listFocusNode[_currentIndex]);
    }
  }

  List<Widget> _buildListWidget() {
    List<Widget> listWidget = [];
    for (int index = 0; index < widget.length; index++) {
      double left = (index == 0) ? 0.0 : (widget.itemSize / 10);
      listWidget.add(Padding(
        padding: widget.margin,
        child: Container(
            height: widget.itemSize,
            width: widget.itemSize,
            margin: EdgeInsets.only(left: left),
            child: _buildInputItem(index)),
      ));
    }
    return listWidget;
  }

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
        scrollDirection: Axis.horizontal,
        child: Column(
          children: <Widget>[
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: _buildListWidget(),
            ),
            widget.clearAll != null
                ? _clearAllWidget(widget.clearAll)
                : const SizedBox(),
          ],
        ));
  }

  void _pasteFromClipboard() async {
    final txt = await _fromClipboard();
    final int txtLength = txt.length;

    widget.onEditing(true);
    for (var i = 0; i < widget.length; i++) {
      _listControllerText[i].text = txtLength > i ? txt[i] : '';
    }
    setState(() {
      _currentIndex = widget.length - 1;
      if (txtLength >= widget.length) {
        widget.onCompleted(_getInputVerify());
      }
      FocusScope.of(context).requestFocus(_listFocusNode[widget.length - 1]);
    });
    widget.onEditing(false);
  }

  Future<String> _fromClipboard() async {
    ClipboardData? cdata = await Clipboard.getData(Clipboard.kTextPlain);
    return cdata?.text ?? '';
  }

  Widget _clearAllWidget(child) {
    return GestureDetector(
      behavior: HitTestBehavior.opaque,
      onTap: () {
        widget.onEditing(true);
        for (var i = 0; i < widget.length; i++) {
          _listControllerText[i].text = '';
        }
        setState(() {
          _currentIndex = 0;
          FocusScope.of(context).requestFocus(_listFocusNode[0]);
        });
        widget.onEditing(false);
      },
      child: child,
    );
  }
}


example/ios/Flutter/flutter_export_environment.sh
#!/bin/sh
# This is a generated file; do not edit or check into version control.
export "FLUTTER_ROOT=/Users/yonatannaor/development/flutter"
export "FLUTTER_APPLICATION_PATH=/Users/yonatannaor/projects/cyber-doron/flutter_verification_code/example"
export "COCOAPODS_PARALLEL_CODE_SIGN=true"
export "FLUTTER_TARGET=/Users/yonatannaor/projects/cyber-doron/flutter_verification_code/example/lib/main.dart"
export "FLUTTER_BUILD_DIR=build"
export "SYMROOT=${SOURCE_ROOT}/../build/ios"
export "FLUTTER_BUILD_NAME=1.0.0"
export "FLUTTER_BUILD_NUMBER=1"
export "DART_DEFINES=Zmx1dHRlci5pbnNwZWN0b3Iuc3RydWN0dXJlZEVycm9ycz10cnVl,RkxVVFRFUl9XRUJfQVVUT19ERVRFQ1Q9dHJ1ZQ=="
export "DART_OBFUSCATION=false"
export "TRACK_WIDGET_CREATION=true"
export "TREE_SHAKE_ICONS=false"
export "PACKAGE_CONFIG=/Users/yonatannaor/projects/cyber-doron/flutter_verification_code/example/.dart_tool/package_config.json"


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_verification_code/flutter_verification_code.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
          primarySwatch: Colors.green,
          primaryColor: Colors.red,
          hintColor: Colors.green),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  bool _onEditing = true;
  String? _code;

  @override
  build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Center(
          child: Text('Example verify code'),
        ),
      ),
      body: Column(
        children: [
          const Padding(
            padding: EdgeInsets.all(8.0),
            child: Center(
              child: Text(
                'Enter your code',
                style: TextStyle(fontSize: 20.0),
              ),
            ),
          ),
          VerificationCode(
            textStyle: Theme.of(context)
                .textTheme
                .bodyText2!
                .copyWith(color: Theme.of(context).primaryColor),
            keyboardType: TextInputType.number,
            underlineColor: Colors
                .amber, // If this is null it will use primaryColor: Colors.red from Theme
            length: 4,
            cursorColor:
                Colors.blue, // If this is null it will default to the ambient
            // clearAll is NOT required, you can delete it
            // takes any widget, so you can implement your design
            clearAll: Padding(
              padding: const EdgeInsets.all(8.0),
              child: Text(
                'clear all',
                style: TextStyle(
                    fontSize: 14.0,
                    decoration: TextDecoration.underline,
                    color: Colors.blue[700]),
              ),
            ),
            margin: const EdgeInsets.all(12),
            onCompleted: (String value) {
              setState(() {
                _code = value;
              });
            },
            onEditing: (bool value) {
              setState(() {
                _onEditing = value;
              });
              if (!_onEditing) FocusScope.of(context).unfocus();
            },
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Center(
              child: _onEditing
                  ? const Text('Please enter full code')
                  : Text('Your code: $_code'),
            ),
          )
        ],
      ),
    );
  }
}
